### PR TITLE
FOGL-6611: Addition of PUT /fledge/security and ACL class

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-6611: Addition of PUT /fledge/security and ACL class

### CODE REVIEW

File: C/common/acl.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+			if (!item.IsObject())
+			{
+				throw new ACLMalformed();
+			}
+			for (Value::ConstMemberIterator itr = item.MemberBegin(); itr != item.MemberEnd(); ++itr)
+			{
+				KeyValueItem i(itr->name.GetString(), itr->value.GetString());
+				m_service.push_back(i);
+
+				//Logger::getLogger()->error("Added %s:%s",
```
Comment (MarkRiddoch) [inline C/common/acl.cpp:110]: Please remove before checkin

File: C/services/common/include/management_api.h [hunk 2]
```diff
@@ -20,8 +20,9 @@
 #define PING			"/fledge/service/ping"
 #define SERVICE_SHUTDOWN	"/fledge/service/shutdown"
 #define CONFIG_CHANGE		"/fledge/change"
-#define CONFIG_CHILD_CREATE "/fledge/child_create"
-#define CONFIG_CHILD_DELETE "/fledge/child_delete"
+#define CONFIG_CHILD_CREATE     "/fledge/child_create"
```
Comment (MarkRiddoch) [inline C/services/common/include/management_api.h:23]: Is this pollution from another branch?

**Review by MarkRiddoch — COMMENTED** [review]
Looks good, just a couple of minor things

**Review by MarkRiddoch — DISMISSED** [review]

**Comment by praveen-garg** [comment]
@deepanshu-yadav This might be of your interest, for ACL configuration item type stuff.

File: C/services/common/include/management_api.h [hunk 3]
```diff
@@ -20,8 +20,9 @@
...
+#define CONFIG_CHILD_DELETE     "/fledge/child_delete"
+#define SECURITY_CHANGE         "^/fledge/security$"
```
Comment (deepanshu-yadav) [inline C/services/common/include/management_api.h:25]: @MarkRiddoch  Should this not be a well known URI, like change endpoint?

File: C/services/common/service_security.cpp [hunk 4]
```diff
@@ -727,3 +743,61 @@ static void bearer_token_refresh_thread(void *data)
...
+	if (r == "attachACL")
+	{
+		// Fetch and load ACL
+		m_service_acl = m_mgtClient->getACL(reason.getArgument());
+
+		// TODO:
+		//  make use of m_service_acl
+		//  in verifyService and verifyURL
+	}
+	else if (r == "reloadACL")
```
Comment (deepanshu-yadav) [inline C/services/common/service_security.cpp:780]: FOGL-6612 says updateACL

File: C/services/common/include/management_api.h (hunk 3 above)
Comment (pintomax) [reply C/services/common/include/management_api.h:25]: @deepanshu-yadav the /fledge/security is a new endpoint for ACL management

Comment (deepanshu-yadav) [reply C/services/common/include/management_api.h:25]: Ok taken that in my pull request.

**Review by MarkRiddoch — APPROVED** [review]

**Review by MarkRiddoch — APPROVED** [review]

**Review by deepanshu-yadav — APPROVED** [review]
Its working now with FOGL-6612.

**Review by deepanshu-yadav — APPROVED** [review]
Working fine...

**Comment by deepanshu-yadav** [comment]
One small issue the service crashes when we attach it to non-existent ACL.

```
curl -skX PUT -H "Content-Type: application/json" -d '{"value": "test1"}' http://localhost:8081/fledge/category/sinSecurity/ACL | jq
{
  "error": {
    "message": "[ServerDisconnectedError] Server disconnected"
  }
}
```

Syslogs

```
Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: ERROR: Failed to fetch ACL: ACL with name test1 is not found..
Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: Signal 6 (Aborted) trapped:
Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: (0) 0       0x55a96d178409 handler(int) + 73---------
Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: (1) 1       0x7f3743676f10 /lib/x86_64-linux-gnu/libc.so.6(+0x3ef10) [0x7f3743676f10]---------
```

**Comment by pintomax** [comment]
> One small issue the service crashes when we attach it to non-existent ACL.
> 
> ```
> curl -skX PUT -H "Content-Type: application/json" -d '{"value": "test1"}' http://localhost:8081/fledge/category/sinSecurity/ACL | jq
> {
>   "error": {
>     "message": "[ServerDisconnectedError] Server disconnected"
>   }
> }
> ```
> 
> Syslogs
> 
> ```
> Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: ERROR: Failed to fetch ACL: ACL with name test1 is not found..
> Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: Signal 6 (Aborted) trapped:
> Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: (0) 0       0x55a96d178409 handler(int) + 73---------
> Aug  1 16:37:09 foglamp-VirtualBox Fledge sin[5944]: FATAL: (1) 1       0x7f3743676f10 /lib/x86_64-linux-gnu/libc.so.6(+0x3ef10) [0x7f3743676f10]---------
> ```

@deepanshu-yadav This is current behaviour of the code: to abort when ACL does not exist.

It can be changed if needed.

**Comment by pintomax** [comment]
Already merged into FOGL-6612

**Pull request closed without merging**